proxy/validation.go

```diff
@@ -0,0 +1,106 @@
+package proxy
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/xeipuuv/gojsonschema"
+)
+
+// Message represents a message flowing through the wire. For example, a Kafka message.
+type Message struct {
+	Context MessageContext
+	Key     []byte
+	Value   []byte
+	Headers []MessageHeader
+}
+
+// MessageContext contains information about the context that surrounds a message.
+type MessageContext struct {
+	Channel string
+}
+
+// MessageHeader represents a header of a message, if there are any.
+type MessageHeader struct {
+	Key   []byte
+	Value []byte
+}
+
+// ValidationError represents a message validation error.
+type ValidationError struct {
+	Msg    *Message
+	Result *gojsonschema.Result
+}
+
+func (v ValidationError) String() string {
+	errs := make([]string, len(v.Result.Errors()))
+	for i, err := range v.Result.Errors() {
+		errs[i] = err.String()
+	}
+
+	return strings.Join(errs, " | ")
+}
+
+// ValidationErrorNotifier notifies whenever a ValidationError happens.
+type ValidationErrorNotifier func(validationError *ValidationError) error
+
+// ValidationErrorToChanNotifier notifies to a given chan when a ValidationError happens.
+func ValidationErrorToChanNotifier(errChan chan *ValidationError) ValidationErrorNotifier {
+	return func(validationError *ValidationError) error {
+		// TODO Blocking or non blocking? Shall we just fire and forget via goroutine instead?
```

Comment:
I don't know 😅

Reply:
TODOS in code are not always a bad thing :)